File: scripts/train/py/inference_lib.py
"""train inference pure functions: threshold application and empty-label fallback.

Shared by evaluate.py (Python) and the TS serving path (src/train/inference-lib.ts) — the two
implementations must behave identically. Light dependency (numpy only).
"""

from typing import Any

import numpy as np
import numpy.typing as npt


def apply_threshold(
    probs: npt.NDArray[np.floating[Any]], threshold: float
) -> npt.NDArray[Any]:
    """sigmoid probabilities -> 0/1 multi-label matrix: at/above the line hits; if a row hits
    nothing, fall back to its highest-scoring class so no empty label set is produced."""
    preds = (probs >= threshold).astype(int)
    for r in range(len(preds)):
        if preds[r].sum() == 0:
            preds[r][probs[r].argmax()] = 1
    return preds

File: src/train/inference-lib.ts
// train inference pure functions: threshold application with an empty-label fallback.
// Shared by the evaluation path and the ONNX serving path — behaviour must not fork.
// No heavy dependencies so the light runtime can import it safely.

export function truncateWithTerminalToken(
  values: number[],
  maxLength: number,
): number[] {
  if (values.length <= maxLength) {
    return values;
  }
  return [...values.slice(0, maxLength - 1), values[values.length - 1]];
}

export function applyThreshold(
  probs: number[][],
  threshold: number,
): number[][] {
  // sigmoid probabilities -> 0/1 multi-label matrix: anything at/above the line hits.
  // If a row hits nothing, fall back to its highest-scoring class so we never emit an empty label set.
  return probs.map((row) => {
    const pred = row.map((p) => (p >= threshold ? 1 : 0));
    if (!pred.some((v) => v === 1)) {
      const argmax = row.indexOf(Math.max(...row));
      pred[argmax] = 1;
    }
    return pred;
  });
}
